File: src/commands/moderation/warnings.ts
import { SlashCommandBuilder, PermissionFlagsBits, EmbedBuilder, ChatInputCommandInteraction } from 'discord.js';
import Warning from '../../models/Warning.js';

export default {
  data: new SlashCommandBuilder()
    .setName('warnings')
    .setDescription('View warnings for a member')
    .setDefaultMemberPermissions(PermissionFlagsBits.ModerateMembers)
    .addUserOption(option =>
      option.setName('target').setDescription('Member to check').setRequired(true)
    )
    .addIntegerOption(option =>
      option.setName('page').setDescription('Page number').setMinValue(1).setRequired(false)
    ),
  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    if (!interaction.guild) {
      await interaction.reply({ content: 'This command can only be used in a server.', ephemeral: true });
      return;
    }

    const user = interaction.options.getUser('target', true);
    const page = (interaction.options.getInteger('page') || 1) - 1;
    const perPage = 5;

    let warnings;
    try {
      warnings = await Warning.find({ guildId: interaction.guild.id, userId: user.id })
        .sort({ createdAt: -1 })
        .skip(page * perPage)
        .limit(perPage)
        .lean();
    } catch (_) {
      await interaction.reply({ content: 'Database is unavailable. Warnings cannot be retrieved.', ephemeral: true });
      return;
    }

    const total = await Warning.countDocuments({ guildId: interaction.guild.id, userId: user.id }).catch(() => 0);

    if (total === 0) {
      await interaction.reply({ content: `${user.tag} has no warnings in this server.`, ephemeral: true });
      return;
    }

    const embed = new EmbedBuilder()
      .setTitle(`⚠️ Warnings for ${user.tag}`)
      .setColor('#f59e0b')
      .setThumbnail(user.displayAvatarURL({ size: 128 }))
      .setFooter({ text: `Total: ${total} warning${total !== 1 ? 's' : ''} | Page ${page + 1} of ${Math.ceil(total / perPage)}` });

    if (warnings.length === 0) {
      embed.setDescription('No warnings on this page.');
    } else {
      warnings.forEach((w, i) => {
        embed.addFields({
          name: `#${page * perPage + i + 1} — <t:${Math.floor(new Date(w.createdAt).getTime() / 1000)}:R>`,
          value: `**Reason:** ${w.reason}\n**Moderator:** <@${w.moderatorId}>`
        });
      });
    }

    await interaction.reply({ embeds: [embed], ephemeral: true });
  }
};

File: src/bot/commands/moderation/warnings.rs
use crate::bot::{Context, Error};
use crate::db::models::{bson_dt_to_chrono, Warning};
use poise::serenity_prelude as serenity;

const PER_PAGE: u64 = 5;

/// View warnings for a member
#[poise::command(
    slash_command,
    ephemeral,
    guild_only,
    required_permissions = "MODERATE_MEMBERS"
)]
pub async fn warnings(
    ctx: Context<'_>,
    #[description = "Member to check"] user: serenity::User,
    #[description = "Page number"]
    #[min = 1_u32]
    page: Option<u32>,
) -> Result<(), Error> {
    let guild_id = ctx
        .guild_id()
        .ok_or_else(|| anyhow::anyhow!("Not in a guild"))?
        .to_string();
    let Some(db) = ctx.data().database() else {
        ctx.say("Database is unavailable. Warnings cannot be retrieved.")
            .await?;
        return Ok(());
    };

    let page = page.unwrap_or(1).saturating_sub(1) as u64;
    let total = Warning::count(&db, &guild_id, &user.id.to_string()).await?;

    if total == 0 {
        ctx.say(format!("{} has no warnings in this server.", user.name))
            .await?;
        return Ok(());
    }

    let warnings = Warning::find_paginated(
        &db,
        &guild_id,
        &user.id.to_string(),
        page * PER_PAGE,
        PER_PAGE as i64,
    )
    .await?;

    let total_pages = total.div_ceil(PER_PAGE);
    let mut embed = serenity::CreateEmbed::new()
        .title(format!("⚠️ Warnings for {}", user.name))
        .colour(serenity::Colour::from_rgb(245, 158, 11))
        .thumbnail(user.face())
        .footer(serenity::CreateEmbedFooter::new(format!(
            "Total: {total} warning{} | Page {} of {total_pages}",
            if total != 1 { "s" } else { "" },
            page + 1
        )));

    if warnings.is_empty() {
        embed = embed.description("No warnings on this page.");
    } else {
        for (i, w) in warnings.iter().enumerate() {
            let ts = bson_dt_to_chrono(w.created_at).timestamp();
            embed = embed.field(
                format!("#{} — <t:{ts}:R>", page * PER_PAGE + i as u64 + 1),
                format!(
                    "**Reason:** {}\n**Moderator:** <@{}>",
                    w.reason, w.moderator_id
                ),
                false,
            );
        }
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}
